File: nxflix-agents-ts/src/models/quiz.ts
import { z } from 'zod';
import { ContentType } from './content-type.js';

/**
 * Question types for all JLPT N1 content.
 */
export const QuestionType = z.enum([
  // General types
  'multiple_choice',
  'fill_in_blank',
  'translation',
  'sentence_construction',
  'error_correction',
  // Kanji-specific types
  'kanji_reading',      // Read the kanji (select correct reading)
  'kanji_meaning',      // Select the meaning of kanji
  'kanji_compound',     // Complete compound word with kanji
  'kanji_write',        // Write the kanji from reading
  // Vocabulary-specific types
  'vocab_meaning',      // Select meaning from word
  'vocab_reading',      // Select reading from word
  'vocab_usage',        // Select correct usage in context
  'vocab_synonym',      // Select synonymous word
  // Listening-specific types
  'listening_comprehension', // Answer from audio
  'listening_task',          // Complete task from audio
  'listening_detail',        // Answer detail question
  // Reading-specific types
  'reading_comprehension',   // Standard MC from passage
  'reading_inference',       // Infer meaning/intent
  'reading_vocabulary',      // Vocabulary question in context
]);
export type QuestionType = z.infer<typeof QuestionType>;

export const QuizQuestion = z.object({
  id: z.string(),
  itemId: z.string(),
  contentType: ContentType.default('grammar'),
  questionType: QuestionType,
  questionText: z.string(),
  questionTextJp: z.string().nullish(),
  options: z.array(z.string()).nullish(),
  correctAnswer: z.string(),
  correctOptionIndex: z.number().int().min(0).max(3).nullish(),
  explanation: z.string(),
  difficulty: z.number().int().min(1).max(5).default(3),
  hints: z.array(z.string()).default([]),
  // For listening questions
  audioUrl: z.string().nullish(),
  // For reading questions
  passageId: z.string().nullish(),
});
export type QuizQuestion = z.infer<typeof QuizQuestion>;

export const Quiz = z.object({
  id: z.string(),
  userId: z.string(),
  itemIds: z.array(z.string()),
  contentTypes: z.array(ContentType).default([]),
  questions: z.array(QuizQuestion),
  difficulty: z.number().int().min(1).max(5).default(3),
  timeLimitSeconds: z.number().int().nullish(),
  createdAt: z.string().datetime().optional(),
});
export type Quiz = z.infer<typeof Quiz>;

export const QuizAnswer = z.object({
  questionId: z.string(),
  userAnswer: z.string(),
  timeTakenSeconds: z.number().nullish(),
});
export type QuizAnswer = z.infer<typeof QuizAnswer>;

export const GradedAnswer = z.object({
  questionId: z.string(),
  userAnswer: z.string(),
  correctAnswer: z.string(),
  isCorrect: z.boolean(),
  score: z.number().min(0).max(1),
  feedback: z.string(),
  grammarExplanation: z.string().nullish(),
});
export type GradedAnswer = z.infer<typeof GradedAnswer>;

export const QuizSubmission = z.object({
  quizId: z.string(),
  answers: z.array(QuizAnswer),
  totalTimeSeconds: z.number().nullish(),
});
export type QuizSubmission = z.infer<typeof QuizSubmission>;

File: nxflix-agents-python/src/nxflix_agents/models/quiz.py
"""Quiz and question models."""

from enum import Enum
from pydantic import BaseModel, Field


class QuestionType(str, Enum):
    """Types of quiz questions."""

    MULTIPLE_CHOICE = "multiple_choice"
    FILL_IN_BLANK = "fill_in_blank"
    TRANSLATION = "translation"
    SENTENCE_CONSTRUCTION = "sentence_construction"
    ERROR_CORRECTION = "error_correction"


class QuizQuestion(BaseModel):
    """A single quiz question."""

    id: str
    grammar_id: str
    question_type: QuestionType
    question_text: str
    question_text_jp: str | None = None
    options: list[str] | None = None  # For multiple choice
    correct_answer: str
    explanation: str
    difficulty: int = Field(default=3, ge=1, le=5)  # 1=easy, 5=hard
    hints: list[str] = []


class Quiz(BaseModel):
    """A complete quiz with multiple questions."""

    id: str
    user_id: str
    grammar_ids: list[str]
    questions: list[QuizQuestion]
    difficulty: int = Field(default=3, ge=1, le=5)
    time_limit_seconds: int | None = None
    created_at: str | None = None


class QuizAnswer(BaseModel):
    """User's answer to a quiz question."""

    question_id: str
    user_answer: str
    time_taken_seconds: float | None = None


class GradedAnswer(BaseModel):
    """A graded answer with feedback."""

    question_id: str
    user_answer: str
    correct_answer: str
    is_correct: bool
    score: float = Field(ge=0.0, le=1.0)  # 0-1 partial credit
    feedback: str
    grammar_explanation: str | None = None


class QuizSubmission(BaseModel):
    """Complete quiz submission with all answers."""

    quiz_id: str
    answers: list[QuizAnswer]
    total_time_seconds: float | None = None
